src/omerocrate/uploader.py

```diff
@@ -314,3 +315,108 @@ async def upload_images(self, image_paths: list[Path], dataset: gateway.DatasetW
                         # Add the image to the dataset
                         dataset._linkObject(wrapper, "DatasetImageLinkI")
                         yield wrapper
+
+
+class SegmentationUploader(ApiUploader):
+    """
+    Subclass of OmeroUploader that also uploads segmentation masks.
+    This class is just a temporary prototype to get this feature working.
+    """
+
+    def find_images_with_segmentation(self) -> Iterable[tuple[Identifier, Path, Path]]:
+        """
+        Finds images containing segmentation masks that should be uploaded to OMERO.
+        Can be overridden to customize the query.
+        """
+        for result in self.select_many("""
+            SELECT ?file_path ?segmentation_file
+            WHERE {
+                ?file_path a schema:MediaObject ;
+                    omerocrate:upload true ;
+                    omerocrate:segmentationFor ?segmentation_file .
+            }
+        """):
+            file_path = result['file_path']
+            segmentation_file = result['segmentation_file']
+            yield file_path, Path(urlparse(file_path).path), Path(urlparse(segmentation_file).path)
+
+    def process_segmentation(self, segmentation_path: Path, image: gateway.ImageWrapper) -> None:
+        """
+        Load segmentation mask and upload to OMERO for the given image URI.
+        By default, uses Glencoe's ROI_Converter_NGFF to parse file, rasterise shapes
+        into a zarr file, and register the mask with OMERO.
+        Can be overridden to utilise an alternative segmentation upload method.
+        """
+        header: list[str] = []
+        with open(segmentation_path, 'r') as f:
+            header = f.readline().strip().split(',')
+
+        # Check geometry column name
+        geometry_column = None
+        if "polygon" in header:
+            geometry_column = "polygon"
+        elif "geometry" in header:
+            geometry_column = "geometry"
+        if not geometry_column:
+            raise ValueError(
+                f"Segmentation file {segmentation_path} does not contain a "
+                f"'{geometry_column}' or 'polygon' column"
+            )
+
+        from ROI_Converter_NGFF import raster
+        args = {
+            "input_file": str(segmentation_path),
+            "register_to": image.getId(),
+            "name": segmentation_path.stem,
+            "directory": os.getenv("UPLOAD_DIRECTORY", None),
+            "output_filename": None,
+            "width": None,
+            "height": None,
+            "tile_size": 2048,
+            "no_fill": False,
+            "server": os.getenv("OMERO_HOST", "localhost"),
+            "port": os.getenv("OMERO_PORT", "4064"),
+            "user": self.conn.getUser().getName(),
+            "password": os.getenv("OMERO_PASSWORD", ""),
```

Comment:
I would rather you use the `conn` to get these variables where possible, because the user may not necessarily be using environment variables to authenticate.

Reply:
Good point, on first glance I couldn't find all the properties in conn, but have resolved this. Needed to use session keys instead of password (as I don't think the latter is possible to return from `conn`), but this ends up being a cleaner way to do it anyway!